### PR TITLE
[closed]

### PR DESCRIPTION
**Pull request by CYarros10**
refer to https://github.com/apache/airflow/pull/37721
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by potiuk — APPROVED** [review]
LGTM. @MaksYermak @VladaZakharova - wdyt?

File: airflow/providers/google/cloud/hooks/vertex_ai/generative_model.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""This module contains a Google Cloud Vertex AI Generative Model hook."""
+
+from __future__ import annotations
+
+from typing import Sequence
+
+import vertexai
```
Comment (VladaZakharova) [inline airflow/providers/google/cloud/hooks/vertex_ai/generative_model.py:24]: Can you please tell what is this package? I can't find it in providers.yaml file and i think for vertexAI we are using google-cloud-aiplatform, no?

File: tests/system/providers/google/cloud/vertex_ai/example_vertex_ai_generative_model.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    description="Sample DAG with generative models.",
+    schedule="@once",
+    start_date=datetime(2024, 1, 1),
+    catchup=False,
+    tags=["example", "vertex_ai", "generative_model"],
+) as dag:
+    # [START how_to_cloud_vertex_ai_prompt_language_model_operator]
+    prompt_language_model_task = PromptLanguageModelOperator(
+        task_id="prompt_language_model_task",
+        project_id="your-project",
```
Comment (VladaZakharova) [inline tests/system/providers/google/cloud/vertex_ai/example_vertex_ai_generative_model.py]: Please, check examples of other system tests in google provider. Specifically, how we set value for project_id and env_id

File: airflow/providers/google/cloud/operators/vertex_ai/generative_model.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+from airflow.providers.google.cloud.hooks.vertex_ai.generative_model import GenerativeModelHook
+from airflow.providers.google.cloud.operators.cloud_base import GoogleCloudBaseOperator
+
+if TYPE_CHECKING:
+    from vertexai.preview.generative_models import ChatSession
+
+    from airflow.utils.context import Context
+
+
+class PromptLanguageModelOperator(GoogleCloudBaseOperator):
```
Comment (VladaZakharova) [inline airflow/providers/google/cloud/operators/vertex_ai/generative_model.py:33]: Can you please describe what is the purpose of having this as a separate operator in separate module? I think we have almost the same functionality in model_service.py so seems like we can extend it?

Comment (CYarros10) [reply airflow/providers/google/cloud/operators/vertex_ai/generative_model.py:33]: Great question! My interpretation is that the ModelServiceClient (used by ModelServiceHook) is for models that the user is creating, deploying, and maintaining.  Whereas the operators in this PR are models that are directly serviced by Google.  Say for example I use the ModelServiceClient to list models in my Google Cloud Project, the generative models such as gemini-pro, text-bison, etc. will not be accessible.

The ModelServiceHook uses a separate python library `from google.cloud.aiplatform_v1.types` whereas GenerativeModelHook currently uses `from vertexai`

There seemed to be separation between the two use cases so I opted to separate for now, with potentially merging down the line when/if Google unifies them into one package.

File: airflow/providers/google/cloud/hooks/vertex_ai/generative_model.py (hunk 1 above)
Comment (CYarros10) [reply airflow/providers/google/cloud/hooks/vertex_ai/generative_model.py:24]: Yes! It is somewhat hidden from the user, but contained in `google-cloud-aiplatform` [here.](https://pypi.org/project/google-cloud-aiplatform/) You can view the source code in the repo: `googleapis/python-aiplatform` [here.](https://github.com/googleapis/python-aiplatform/tree/main/vertexai) They discuss importing in the [README](https://github.com/googleapis/python-aiplatform?tab=readme-ov-file#importing).

File: tests/system/providers/google/cloud/vertex_ai/example_vertex_ai_generative_model.py (hunk 2 above)
Comment (CYarros10) [reply tests/system/providers/google/cloud/vertex_ai/example_vertex_ai_generative_model.py]: This is good feedback, I will address in the next commit. Thank you.

File: tests/system/providers/google/cloud/vertex_ai/example_vertex_ai_generative_model.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    )
+    # [END how_to_cloud_vertex_ai_prompt_language_model_operator]
+
+    # [START how_to_cloud_vertex_ai_prompt_multimodal_model_operator]
+    prompt_multimodal_model_task = PromptMultimodalModelOperator(
+        task_id="generative_model_task",
+        project_id=PROJECT_ID,
+        location=REGION,
+        prompt=PROMPT,
+        pretrained_model=MULTIMODAL_MODAL,
```
Comment (VladaZakharova) [inline tests/system/providers/google/cloud/vertex_ai/example_vertex_ai_generative_model.py]: I think it should be `MULTIMODAL_MODEL`? :)
Please, correct me if i am wrong :D

File: airflow/providers/google/cloud/hooks/vertex_ai/generative_model.py (hunk 1 above)
Comment (VladaZakharova) [reply airflow/providers/google/cloud/hooks/vertex_ai/generative_model.py:24]: Got it!
Thank you for explanation :)
So, this `vertexai` is available only for google-cloud-aiplatform>=1.41.0, right?

File: tests/system/providers/google/cloud/vertex_ai/example_vertex_ai_generative_model.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    )
+    # [END how_to_cloud_vertex_ai_prompt_multimodal_model_operator]
+
```
Comment (ahidalgob) [inline tests/system/providers/google/cloud/vertex_ai/example_vertex_ai_generative_model.py:78]:
```suggestion
    from tests.system.utils.watcher import watcher

    # This test needs watcher in order to properly mark success/failure
    # when "tearDown" task with trigger rule is part of the DAG
    list(dag.tasks) >> watcher()
```

File: airflow/providers/google/cloud/hooks/vertex_ai/generative_model.py (hunk 1 above)
Comment (CYarros10) [reply airflow/providers/google/cloud/hooks/vertex_ai/generative_model.py:24]: Correct. somewhat confusing, there is an outdated (May 2023) pypi package [vertexai](https://pypi.org/project/vertexai/), but the project description indicates: To use the Vertex AI SDK, please install the google-cloud-aiplatform PyPi package by running pip install google-cloud-aiplatform. So this google-cloud-aiplatform package is being used here

**Pull request closed without merging**